File: src/glpk/examples/vb/gmpl.vb
Imports System
Imports org.gnu.glpk

Module module1
	Function Main(ByVal cmdArgs() As String) As Integer
	    Dim problem As clsGmpl
	    
		If 1 <> cmdArgs.Length Then
			Console.WriteLine ("Usage: gmpl.exe model.mod")
			Return 1
		End If
		problem = new clsGmpl ()
		Problem.solve (cmdArgs)
	End Function
End Module

Public Class clsGmpl
	Implements IGlpkCallbackListener, IGlpkTerminalListener

	Private hookUsed As Boolean = false

	Public Sub New()
	End Sub

	Sub solve (arg() As string)
		Dim lp As glp_prob
		Dim tran As glp_tran
		Dim iocp As glp_iocp

		Dim fname As String
		Dim skip As Integer = 0
		Dim ret As Integer

		GLPK.glp_cli_set_numeric_locale ("C")

		' listen to callbacks
		GlpkCallback.addListener (Me)

		' listen to terminal output
		GlpkTerminal.addListener (Me)

		Try

			' create problem
			lp = GLPK.glp_create_prob ()

			' allocate workspace
			tran = GLPK.glp_mpl_alloc_wksp ()

			' read model
			fname = arg(0)
			ret = GLPK.glp_mpl_read_model (tran, fname, skip)
			If ret <> 0 Then
				GLPK.glp_mpl_free_wksp (tran)
				GLPK.glp_delete_prob (lp)
				throw new ApplicationException ( _
					"Model file not valid: " + fname)
			End If

			' generate model
			ret = GLPK.glp_mpl_generate (tran, Nothing)
			If ret <> 0 Then
				GLPK.glp_mpl_free_wksp (tran)
				GLPK.glp_delete_prob (lp)
				throw new ApplicationException ( _
					"Cannot generate model: " + fname)
			End If

			' build model
			GLPK.glp_mpl_build_prob (tran, lp)

			' set solver parameters
			iocp = new glp_iocp ()
			GLPK.glp_init_iocp (iocp)
			iocp.presolve = GLPK.GLP_ON

			' do not listen to output anymore
			GlpkTerminal.removeListener (Me)

			' solve model
			ret = GLPK.glp_intopt (lp, iocp)

			' postsolve model
			If ret = 0 Then
				GLPK.glp_mpl_postsolve (tran, lp, GLPK.GLP_MIP)
			End If

			' free memory
			GLPK.glp_mpl_free_wksp (tran)
			GLPK.glp_delete_prob (lp)

		Catch e As org.gnu.glpk.GlpkException
			Console.Error.WriteLine ( _
				"An error inside the GLPK library occured.")
			Console.Error.WriteLine (e.Message)
		Catch e As ApplicationException
			Console.Error.WriteLine ( _
				"Applicaiton Exception")
			Console.Error.WriteLine (e.Message)
		End Try

		' do not listen for callbacks anymore
		GlpkCallback.removeListener (Me)

		' check that the terinal hook function has been used
		If Not hookUsed Then
			throw new ApplicationException ( _
				"The terminal output hook was not used.")
		End If
	End Sub

	Public Function output (str As string) As Boolean _
	    Implements IGlpkTerminalListener.output

		hookUsed = true
		Console.Write (str)
		return false
	End Function

	Public Sub callback (tree As glp_tree) _
	    Implements IGlpkCallbackListener.callback
	    
	    Dim reason as Integer
		reason = GLPK.glp_ios_reason (tree)
		If reason = GLPK.GLP_IBINGO Then
			Console.WriteLine ("Better solution found")
		End If
	End Sub
End Class

File: src/glpk/examples/csharp/gmpl.cs
using System;
using org.gnu.glpk;

public class Gmpl : IGlpkCallbackListener, IGlpkTerminalListener
{
	private bool hookUsed = false;

	public static void Main (string[] arg)
	{
		if (1 != arg.Length) {
			Console.WriteLine ("Usage: gmpl.exe model.mod");
			return;
		}
		GLPK.glp_cli_set_numeric_locale ("C");
		new Gmpl ().solve (arg);
	}

	public void solve (string[] arg)
	{
		glp_prob lp;
		glp_tran tran;
		glp_iocp iocp;

		String fname;
		int skip = 0;
		int ret;

		// listen to callbacks
		GlpkCallback.addListener (this);

		// listen to terminal output
		GlpkTerminal.addListener (this);

		try {

			// create problem
			lp = GLPK.glp_create_prob ();

			// allocate workspace
			tran = GLPK.glp_mpl_alloc_wksp ();

			// read model
			fname = arg [0];
			ret = GLPK.glp_mpl_read_model (tran, fname, skip);
			if (ret != 0) {
				GLPK.glp_mpl_free_wksp (tran);
				GLPK.glp_delete_prob (lp);
				throw new ApplicationException ("Model file not valid: " + fname);
			}

			// generate model
			ret = GLPK.glp_mpl_generate (tran, null);
			if (ret != 0) {
				GLPK.glp_mpl_free_wksp (tran);
				GLPK.glp_delete_prob (lp);
				throw new ApplicationException ("Cannot generate model: " + fname);
			}

			// build model
			GLPK.glp_mpl_build_prob (tran, lp);

			// set solver parameters
			iocp = new glp_iocp ();
			GLPK.glp_init_iocp (iocp);
			iocp.presolve = GLPK.GLP_ON;

			// do not listen to output anymore
			GlpkTerminal.removeListener (this);

			// solve model
			ret = GLPK.glp_intopt (lp, iocp);

			// postsolve model
			if (ret == 0) {
				GLPK.glp_mpl_postsolve (tran, lp, GLPK.GLP_MIP);
			}

			// free memory
			GLPK.glp_mpl_free_wksp (tran);
			GLPK.glp_delete_prob (lp);

		} catch (org.gnu.glpk.GlpkException e) {
			Console.Error.WriteLine ("An error inside the GLPK library occured.");
			Console.Error.WriteLine (e.Message);
		} catch (ApplicationException e) {
			Console.Error.WriteLine (e.Message);
		}

		// do not listen for callbacks anymore
		GlpkCallback.removeListener (this);

		// check that the terinal hook function has been used
		if (!hookUsed) {
			throw new ApplicationException (
				"The terminal output hook was not used.");
		}
	}

	public bool output (string str)
	{
		hookUsed = true;
		Console.Write (str);
		return false;
	}

	public void callback (glp_tree tree)
	{
		int reason = GLPK.glp_ios_reason (tree);
		if (reason == GLPK.GLP_IBINGO) {
			Console.WriteLine ("Better solution found");
		}
	}
}
